lgc/state/PipelineState.cpp

```diff
@@ -567,36 +576,43 @@ const ResourceNode *PipelineState::findPushConstantResourceNode() const {
 // @param binding : ID of descriptor binding
 std::pair<const ResourceNode *, const ResourceNode *>
 PipelineState::findResourceNode(ResourceNodeType nodeType, unsigned descSet, unsigned binding) const {
+  // Implements the <= order for the ResourceNodeType lattice.
+  // Returns true iff nodeType <= candidateType.
+  auto IsNodeTypeCompatible = [nodeType](ResourceNodeType candidateType) {
+    if (nodeType == ResourceNodeType::Unknown || candidateType == nodeType)
+      return true;
+
+    if (nodeType == ResourceNodeType::DescriptorBuffer &&
+        (candidateType == ResourceNodeType::DescriptorBufferCompact || candidateType == ResourceNodeType::InlineBuffer))
+      return true;
+
+    if ((nodeType == ResourceNodeType::DescriptorResource || nodeType == ResourceNodeType::DescriptorTexelBuffer ||
+         nodeType == ResourceNodeType::DescriptorSampler) &&
+        (candidateType == ResourceNodeType::DescriptorCombinedTexture ||
+         candidateType == ResourceNodeType::DescriptorYCbCrSampler))
+      return true;
+
+    return false;
+  };
+
   for (const ResourceNode &node : getUserDataNodes()) {
     if (node.type == ResourceNodeType::DescriptorTableVaPtr) {
-      if (nodeType == node.type) {
-        if (!node.innerTable.empty() && node.innerTable[0].set == descSet)
+      // Check inner nodes.
+      bool isFirstChild = true;
+      for (const ResourceNode &innerNode : node.innerTable) {
+        if (isFirstChild && nodeType == ResourceNodeType::DescriptorTableVaPtr && innerNode.set == descSet)
           return {&node, &node};
-      } else {
-        for (const ResourceNode &innerNode : node.innerTable) {
-          if (innerNode.set == descSet && innerNode.binding == binding) {
-            if (nodeType == ResourceNodeType::Unknown || nodeType == innerNode.type ||
-                (nodeType == ResourceNodeType::DescriptorBuffer &&
-                 (innerNode.type == ResourceNodeType::DescriptorBufferCompact ||
-                  innerNode.type == ResourceNodeType::InlineBuffer)) ||
-                ((innerNode.type == ResourceNodeType::DescriptorCombinedTexture ||
-                  innerNode.type == ResourceNodeType::DescriptorYCbCrSampler) &&
-                 (nodeType == ResourceNodeType::DescriptorResource ||
-                  nodeType == ResourceNodeType::DescriptorTexelBuffer ||
-                  nodeType == ResourceNodeType::DescriptorSampler)))
-              return {&node, &innerNode};
-          }
-        }
+        isFirstChild = false;
```

Comment:
If `nodeType == DescriptorTableVaPtr`, then there's no reason to ever iterate over the innerTable, is there? The old code here was clearer in that regard..

Reply:
I agree.  There should not be nested `ResourceNodeType::DescriptorTableVaPtr` nodes.  If you are looking for a `ResourceNodeType::DescriptorTableVaPtr` node and the first inner node is not in the correct descriptor set, then there is no need to check the rest of the nodes.

Reply:
I reverted to the old check, as you find it more clean. However, putting the special case in the first loop iteration should have been logically the same.

Reply:
Edit: I mismatched braces in the original code and see what you meant now.